feat(progress): status api, 'statusline' integration

shadmansaleh opened the pull request:
Follow up to #34846

### Problem
Currently, statusline by default doesn't show progress status.

### Solution
Add progress-message support to the default statusline.

### How it works
Displays statusline text summarizing 'running' progress messages.
 - If none: returns empty string
 - If one running item: "title:  percent%"
 - If multiple running items: "Progress: N items avg-percent%"


### Examples
<img width="1920" height="27" alt="image" src="https://github.com/user-attachments/assets/799716cd-7d27-45ef-b642-a8b492d970a1" />
<img width="1920" height="27" alt="image" src="https://github.com/user-attachments/assets/88a86562-5cd9-40b3-82a5-fbccc4dfb618" />

shadmansaleh committed:
feat: add statusline integration for progress messages

shadmansaleh commented on src/nvim/options.lua:
```diff
@@ -8729,6 +8729,7 @@ local options = {
           '%<',
           '%f %h%w%m%r ',
           '%=',
+          "%{% luaeval(\"package.loaded['vim.status'] and vim.status.get_progress_status() or ''\") %}",
```
Note(self): This should fix the test error in channel_spec. However, now since by default vim. status is lazy loaded it doesn't get activated at all by default. Need to figure something out. vim.status should be lazy-loaded. But now the question is what triggers it to activate.

shadmansaleh replied:
@justinmk anyidea what can be done for this? Reading the conversation on the diagnostinc-status pr and it's followups this seems to be a limiation in our current module system.

shadmansaleh commented on runtime/lua/vim/status.lua:
```diff
@@ -0,0 +1,71 @@
+        or ev.data.status == 'failed'
+        or ev.data.status == 'cancel'
+      then
+        progress[ev.data.msg_id] = nil
```
> we have vim.ringbuf which can at least set a maximum (could be very high)

If I understand right, `vim.ringbuf` doesn't feel suitable. it's a queue with a max length set at creation, right? We do need map since any item can expire at any time, so the order we remove the items isn't linear. Also,  random access through indexing is also useful for us. 

> if the job is stuck then that implies we also are storing state for the job itself, right? and that's more expensive than this message data.

True, although I wasn't worried about memory uses exactly more like UI getting stale.

shadmansaleh commented:
> vim.ui seems appropriate, what is the motivation for a new vim.status module ?

The motivation was to keep all statusline-related stuff together somewhere. But I think using vim.ui is fine.

> Also check if https://github.com/neovim/neovim/issues/35152#issuecomment-3205352525 has any implications here.

I feel like adding new statusline items adds unnecessary complications. Since it basically involves dealing with the statusline-expression parser. Also if we keep adding new items like that we will run out of letters or will have to settle for suboptimal abbebiation. For example, %p taken so we can't really say p for progress anymore. Using Lua functions makes adding things to statusline significantly easier. If anything, we should provide a better way to run Lua expressions from the statusline.

Populating global/local variables and building a statusline from that isn't bad idea in my opinion. Before neovim and lua got endorsed vim stuff used to work like that. That way still works. And because executing VimL expressions in the statusline is supported and viml can conviniently access those variables statusline integration in that way becomes much simpler. 

I was actually testing setting g:progress_status variable on the event callback. It basically works almost the same.

justinmk commented:
> The motivation was to keep all statusline-related stuff together somewhere. But I think using vim.ui is fine.

Not sure. I would lean towards putting it in `vim.ui` until we have a clear reason not to. Maybe we need to introduce `vim.msg` for message-related functionality (progress, notifications, ...?).

> I feel like adding new statusline items adds unnecessary complications. Since it basically involves dealing with the statusline-expression parser. Also if we keep adding new items like that we will run out of letter

shadmansaleh commented on runtime/doc/lua.txt:
```diff
@@ -4734,6 +4734,23 @@ vim.text.indent({size}, {text}, {opts})                    *vim.text.indent()*
 ==============================================================================
 Lua module: vim.ui                                                    *vim.ui*
 
+*vim.ui.get_progress_status.Opts*
```
Why does this get generated separately?  Shouldn't this be part of the function doc? like other functions

antonk52 commented on runtime/lua/vim/ui.lua:
```diff
@@ -253,4 +253,95 @@ function M._get_urls()
+  ---Return statusline text summarizing progress messages.
+  --- - If none: returns empty string
+  --- - If one running item: "title: 42%"
+  --- - If multiple running items: "Progress: N items AVG%"
```
Personally, I prefer `{avg-percent}%({count})` as it's immediately clear what the count number relates to. In the screenshot in the PR description it was not immediately obvious what 10% relates to as most numbers in statusline do not have space gaps with their labels if present

shadmansaleh commented on runtime/doc/lua.txt:
```diff
@@ -4734,6 +4734,23 @@ vim.text.indent({size}, {text}, {opts})                    *vim.text.indent()*
+                progress messages
+
+
+vim.ui.get_progress_status({opts})              *vim.ui.get_progress_status()*
```
I think `ui.progress_status` is better. This is statusline specific. We can leave `ui.progress` for future use if we need it.

justinmk reviewed (approved):
Seems like a fine starting point. Would be nice to have this rebased and merged in 0.12

This also would pave the way for https://github.com/neovim/neovim/issues/36209

shadmansaleh committed:
refactor(statusline): add way to get managed active progress items

- removes fmt option for progress_status
- adds vim.ui.progress_get_running

shadmansaleh committed:
address review: return running progress messages as 2nd value of
progress_status

shadmansaleh committed:
address review: 2

- Create event listeners on first call to progress_status() so any one
  not using it doesn't get any penalty for it
- Dropped list of running progress item as second return. Custom setups
  can directly handle it through Progress event

shadmansaleh committed:
run ci

justinmk committed:
Apply suggestion from justinmk

justinmk committed:
Apply suggestion from justinmk

justinmk commented on src/nvim/options.lua:
```diff
@@ -8795,6 +8795,7 @@ local options = {
           '%f %h%w%m%r ',
           "%{% v:lua.require('vim._core.util').term_exitcode() %}",
           '%=',
+          "%{% luaeval('(package.loaded[''vim.ui''] and vim.ui.progress_status()) or '''' ')%}",
```
I'm hesitating on this because showing progress, which is a "global" concept, in every window's statusline, seems strange. Long term, we should think about how it could be surfaced in ui2 (cmdline area), or maybe in the default tabline (top right).

We may want to revert this before releasing 0.12. We want the default statusline to be something users build on, not something they have to surgically revert. So its choices need to be solid.
